### ISSUE TITLE
Unexpected behavior when using new data-driven tests with non-data classes

### ISSUE THREAD

**Issue by lwasyl** [kotest/kotest#1795]
Kotest 4.3.0

Given the following test:
```kotlin
package com.example

import io.kotest.core.datatest.forAll
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.ints.shouldBeLessThanOrEqual

data class Working(val data: Int)

class NotWorking(val data: Int)

class TestSpec : DescribeSpec({
    describe("DataSpec") {
        describe("working") {
            forAll(
                Working(1),
                Working(2),
            ) { d ->
                d.data shouldBeLessThanOrEqual 1
            }
        }
        describe("not-working") {
            forAll(
                NotWorking(1),
                NotWorking(2),
            ) { d ->
                d.data shouldBeLessThanOrEqual 1
            }
        }
    }
}) {

    override fun isolationMode() = IsolationMode.InstancePerLeaf
}
```

It behaves unexpectedly in the not-working block: it only runs the first test case, and completely ignores the following ones:
![image](https://user-images.githubusercontent.com/3951580/96735632-0b579280-13bc-11eb-95db-5a6772be59b4.png)

While I understand this isn't how to use the data-driven tests (we should use data classes), I think Kotest shouldn't fail silently in that case. The failure is less obvious when using `data class TestData(val someValue: Int, val someNonDataClass: Foo)`, where mere having a non-data class as part of the test data triggers the error as well.

I'm also curious what specifically causes this behavior? Is `toString()` for a given data row required to be stable and return the same string? I was playing with returning incrementing integer values in `toString()` and noticed that registering each row case calls `toString()` ~3 times.

**Comment by simonenkoi** [comment]
@sksamuel as per initial debugging, I've found that the second test case is enqueued in this place: https://github.com/kotest/kotest/blob/bcb375f7a7c6ee484832aab7607746ad9d6a7b04/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/runners/InstancePerLeafSpecRunner.kt#L126
A simple removal of the `open` variable with a related conditional branch leads to the expected result:
![image](https://user-images.githubusercontent.com/29431950/96749604-b1fb5d80-13d3-11eb-83ac-e80db131b86c.png)
I couldn't understand this variable's logic, so could you elaborate on this one, please?

**Comment by sksamuel** [comment]
Are you asking me why we call enqueue at that line ?

**Comment by sksamuel** [comment]
Please see https://github.com/kotest/kotest/pull/1798 @lwasyl @simonenkoi

**Comment by sksamuel** [comment]
Will include the fix in 4.3.1
